---
title: 浅谈作用域和闭包问题
date: 2020-03-25
categories:
  - 前端
tags:
  - JavaScript
---

## 一、前言

很多人面试都会被问到什么是闭包，这个问题我也反反复复看过很多次，下面我就来总结一下我自己的见解，也方便自己复习。

## 二、不得不说的作用域

::: tip
个人认为，闭包就是对某个函数传到该词法作用域以外的地方调用（简略的说，可能有错，请各位斧正）。所以应该解释一下作用域是什么
:::

作用域主要有两种工作模型，第一种就是大多数编程语言所采用的词法作用域，另一种叫做动态作用域（有兴趣的同学可以去搜一下）。JavaScript 当然所采用的是词法作用域（以下都统称作用域）。

编译器在开始工作时，第一个阶段叫做词法化，就是将代码中的字符进行检查，如果时有状态的解析过程，还会赋予单词语义。（这也是叫词法作用域的原因）总的来说，词法作用域是由你写代码时将变量和块作用域写在哪里来决定的。例：

```javascript
// A
function foo(a) {
  // B
  var b = a * 2;

  function bar(c) {
    // C
    console.log(a, b, c);
  }

  bar(b * 3);
}

foo(2); // 2, 4, 12
```

以上就包括了 3 个作用域

- A 为最外层，包含整个全局作用域，其中只有标识符 `foo`
- B 为函数 `foo` 所创建的作用域，其中包含 `a, b, bar` 三个标识符
- C 为 `bar` 所创建的作用域， 其中只有 `c` 标识符

每一个函数都会创建一个新的作用域，它是逐级包含的，外层作用域的变量，在里层也可以调用。但是在外层调用不了里层的变量和函数，比如在 A 作用域就调用不了`bar`函数。

当然除了函数作用域，还有一种叫做块作用域，很多刚开始接触 js 的人认为只要有`{}`包裹的代码就是块作用域（。。。先接触 C 的菜鸡本人开始也是这样认为的），但实际上形成块作用域需要一下三种方法

- `with`
- `try/catch`
- `let`

`with`现在一般很少人用，我这就不细讲，有兴趣可以自己搜一下；`try/catch` 只要接触过编程的人一般都知道，在`catch`分句回创建一个块作用域，传入的异常参数只在`catch`中有效；`let` 本人认为这个关键字改变了在 ES5 之前变量莫名其妙被覆盖的状况，其实也解决了闭包难理解的问题。

`let` 可以将变量绑定到任意作用域中，使用`let`定义变量，就可以认为在`{}`中都是一个块作用域。

```javascript
let a = 2;
if (a) {
  var b = 2;
}
console.log(b); // 2

if (a) {
  let b = 2;
}
console.log(b);
// 报错，b不存在
```

第二个例子报错就是因为，在`{}`中用`let`声明的变量`b`形成了个块作用域，外部调用不了它

## 3、闭包的基础概念

:::tip
在对作用域有一定理解之后，我们就可以对闭包进行学习了。在初学 JavaScript 的时候，每个人都谈虎色变，因为它太难理解了，但是好像又不得不掌握它，当别人问起时也说不清道不明。但实际上，闭包并不是一个特别的东西，我们可能在不知情的时候已经使用了闭包，只是我们并不理解这是闭包。以下观点都是我参照各个大佬的文章和书籍进行理解总结的，如果有错欢迎指正。
:::

